scripts/abins/instruments/directinstrument.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import (absolute_import, division, print_function)
+import math
+from typing import Tuple
+
+import numpy as np
+
+from abins.constants import (FLOAT_TYPE, MILLI_EV_TO_WAVENUMBER,
+                             WAVENUMBER_TO_INVERSE_A)
+from .instrument import Instrument
+from .broadening import broaden_spectrum, prebin_required_schemes
+
+
+class DirectInstrument(Instrument):
+    """Simulated direct-geometry INS
+
+    These instruments are suitable for Abins2D. The energy range is
+    determined by a variable incident energy and the corresponding q limits
+    are determined by an angular range.
+    """
+    def __init__(self,
+                 name: str = 'Direct',
+                 setting: str = '') -> None:
+        self._name = name
+        self._e_init = None
+
+        super().__init__(setting=setting)
+
+    def set_incident_energy(self, e_init: float, units: str = 'cm-1'):
+        """
+        Setter for incident energy.
+
+        :param e_init: new incident energy
+        :param units: units of input e_init: 'cm-1' or 'meV'.
+        """
+        if units == 'cm-1':
+            self._e_init = e_init
+        elif units == 'meV':
+            self._e_init = e_init * MILLI_EV_TO_WAVENUMBER
+        else:
+            raise ValueError("Unknown unit: incident energy should be given "
+                             "in 'cm-1' or 'meV'.")
+
+    def get_incident_energy(self) -> float:
+        return self._e_init
+
+    def get_max_wavenumber(self):
+        return self.get_incident_energy()
+
+    def get_min_wavenumber(self):
+        return 0.
+
+    def get_angle_range(self):
+        raise NotImplementedError()
+
+    def get_abs_q_limits(self, energy: np.ndarray) -> np.ndarray:
+        """Get absolute q range accessible for given energy transfer at set incident energy
+
+        Args:
+            energy: 1-D array of energy transfer values
+
+
+        Returns:
+            limits: (2, N) array of q-limits corresponding to energy series
+        """
+        assert len(energy.shape) == 1
+
+        min_angle, max_angle = self.get_angle_range()
+
+        # Actual limits depend on |2θ|, so some shuffling is necessary to get
+        # absolute values in the right order, spanning the whole range.
+        if (min_angle < 0) and (max_angle <= 0):
+            min_angle, max_angle = -max_angle, -min_angle
+        elif (min_angle < 0):
+            # e.g. if -10 < 2θ < 20, we don't want to set limits 10 < |2θ| < 20
+            # as this would omit contributions from the range -10 < θ < 10
```

Comment:
nice comment

Reply:
It took a while to figure out all these cases :sweat_smile: